### PR TITLE
Rebase/roles

### PR DESCRIPTION
**Pull request by 7ep**

### CODE REVIEW

File: src/main/kotlin/coverosR3z/authentication/persistence/AuthenticationPersistence.kt [hunk 1]
```diff
@@ -47,4 +47,11 @@ class AuthenticationPersistence(private val pmd : PureMemoryDatabase, private va
         return pmd.UserDataAccess().read { users -> users.toSet() }
     }
 
+    override fun addRoleToUser(userName: UserName, role: Roles): User {
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/persistence/AuthenticationPersistence.kt:50]: using the username to get this doesn't feel entirely pure

File: src/main/kotlin/coverosR3z/authentication/persistence/IAuthPersistence.kt [hunk 2]
```diff
@@ -13,4 +13,5 @@ interface IAuthPersistence {
     fun deleteSession(user: User)
     fun getAllSessions(): Set<Session>
     fun getAllUsers(): Set<User>
+    fun addRoleToUser(userName: UserName, role: Roles): User
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/persistence/IAuthPersistence.kt:16]: wondering if this is the best approach - alternates: user id? user?

File: src/main/kotlin/coverosR3z/authentication/types/CurrentUser.kt [hunk 3]
```diff
@@ -3,5 +3,5 @@ package coverosR3z.authentication.types
 /**
  * The user currently logged in and executing commands
  */
-data class CurrentUser(val user : User)
+class CurrentUser(user : User) : User(user.id, user.name, user.hash, user.salt, user.employeeId, user.role)
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/types/CurrentUser.kt:6]: ooohhhh

File: src/main/kotlin/coverosR3z/authentication/types/User.kt [hunk 4]
```diff
@@ -101,12 +137,15 @@ data class User(val id: UserId, val name: UserName, val hash: Hash, val salt: Sa
 
                 val empId = EmployeeId(checkParseToInt(entries[Keys.EMPLOYEE_ID]))
 
+                val role = Roles.valueOf(entries[Keys.ROLE]!!)
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/types/User.kt:140]: double bangs? for shame

File: src/main/kotlin/coverosR3z/authentication/types/User.kt [hunk 5]
```diff
@@ -101,12 +137,15 @@ data class User(val id: UserId, val name: UserName, val hash: Hash, val salt: Sa
...
+
                 User(
                     UserId(id),
                     UserName.make(entries[Keys.NAME]),
                     Hash.make(entries[Keys.HASH]),
                     Salt.make(entries[Keys.SALT]),
-                    empId)
+                    empId,
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/types/User.kt:147]: is this right?

File: src/main/kotlin/coverosR3z/authentication/utility/Authorization.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import coverosR3z.authentication.exceptions.UnpermittedOperationException
+import coverosR3z.authentication.types.CurrentUser
+import coverosR3z.authentication.types.Roles
+
+/**
+ * Code for authorization - a regular user can't do what an admin can do, etc.
+ */
+class RolesChecker(val cu: CurrentUser) : IRolesChecker {
+
+    override fun checkAllowed(vararg roles: Roles) {
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/utility/Authorization.kt:12]: magic!!!!

File: src/main/kotlin/coverosR3z/logging/Logger.kt [hunk 7]
```diff
@@ -43,7 +43,7 @@ class Logger : ILogger {
 
     override fun logAudit(cu : CurrentUser, msg : () -> String) {
         if (logSettings[LogTypes.AUDIT] == true) {
-            loggerPrinter.enqueue { println("${getCurrentMillis()} AUDIT: ${cu.user.name.value}: ${msg()}") }
+            loggerPrinter.enqueue { println("${getCurrentMillis()} AUDIT: ${cu.name.value}: ${msg()}") }
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/logging/Logger.kt:46]: yay! atta boy Matt

File: src/test/kotlin/coverosR3z/authentication/utility/RoleVerificationTests.kt [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.Before
+import org.junit.Test
+
+class RoleVerificationTests {
+
+    private lateinit var authUtils : AuthenticationUtilities
+    private lateinit var ap : FakeAuthPersistence
+
+    @Before
+    fun init() {
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/utility/RoleVerificationTests.kt:17]: this is next to do

**Review by 7ep — COMMENTED** [review]
Looks good, put some comments in

**Review by coveros-matt-taylor — APPROVED** [review]
Looks great I've definitely looked over everything thoroughly yes.

**Pull request merged**